src/main/scala/decider/Z3ProverAPI.scala

```diff
@@ -0,0 +1,439 @@
+      }
+      result
+    } catch {
+      case e: Z3Exception => throw ExternalToolError("Prover", "Z3 error: " + e.getMessage)
+    }
+  }
+
+  protected def assertUsingPushPop(goal: Term): (Boolean, Long) = {
+    endPreamblePhase()
+    push()
+
+    val negatedGoal = ctx.mkNot(termConverter.convert(goal).asInstanceOf[BoolExpr])
+    prover.add(negatedGoal)
+    val startTime = System.currentTimeMillis()
+    val res = prover.check()
+    val endTime = System.currentTimeMillis()
+    val result = res == Status.UNSATISFIABLE
+    pop()
+
+    if (!result) {
+      retrieveAndSaveModel()
+    }
+
+    (result, endTime - startTime)
+  }
+
+  def saturate(data: Option[Config.ProverStateSaturationTimeout]): Unit = {
+    endPreamblePhase()
+    data match {
+      case Some(Config.ProverStateSaturationTimeout(timeout, comment)) => saturate(timeout, comment)
+      case None => /* Don't do anything */
+    }
+  }
+
+  def saturate(timeout: Int, comment: String): Unit = {
+    endPreamblePhase()
+    setTimeout(Some(timeout))
+    prover.check()
+  }
+
+  protected def retrieveAndSaveModel(): Unit = {
+    if (Verifier.config.counterexample.toOption.isDefined) {
+      val model = prover.getModel
+      lastModel = model
+    }
+  }
+
+  protected def assertUsingSoftConstraints(goal: Term): (Boolean, Long) = {
+    endPreamblePhase()
+    val guard = fresh("grd", Nil, sorts.Bool)
+    val guardApp = App(guard, Nil)
+    val goalImplication = Implies(guardApp, goal)
+
+    prover.add(termConverter.convertTerm(goalImplication).asInstanceOf[BoolExpr])
+
+    val startTime = System.currentTimeMillis()
+    val res = prover.check(termConverter.convertTerm(guardApp))
+    val endTime = System.currentTimeMillis()
+    val result = res == Status.UNSATISFIABLE
+    if (!result) {
+      retrieveAndSaveModel()
+    }
+
+    (result, endTime - startTime)
+  }
+
+  def check(timeout: Option[Int] = None): Result = {
+    endPreamblePhase()
+    setTimeout(timeout)
+
+    val res = prover.check()
+
+    res match {
+      case Status.SATISFIABLE => Sat
+      case Status.UNSATISFIABLE => Unsat
+      case Status.UNKNOWN=> Unknown
+    }
+  }
+
+  def endPreamblePhase() =  {
+    if (!preamblePhaseOver) {
+      preamblePhaseOver = true
+
+      val merged = emittedPreambleString.mkString("\n")
+      val parsed = ctx.parseSMTLIB2String(merged, emittedSortSymbols.toArray, emittedSorts.toArray, emittedFuncSymbols.toArray, emittedFuncs.toArray)
+      prover.add(parsed: _*)
+      prover.add(preambleAssumes.toSeq : _*)
+      preambleAssumes.clear()
+    }
+  }
+
+  def statistics(): Map[String, String] = {
+    val statistics = prover.getStatistics
+    val result = mutable.HashMap[String, String]()
+    for (e <- statistics.getEntries()) {
+      result.update(e.Key, e.getValueString)
+    }
+    ListMap.from(result)
+  }
+
+  def comment(str: String): Unit = {
+    if (!preamblePhaseOver && str == "End preamble")
```

Comment:
Is this `str == ...` check ever necessary to detect the end of preamble?

Reply:
Apparently not, I'll remove it.